File: app/src/main/java/com/eh/recyclerlistsample/MainActivity.java
package com.eh.recyclerlistsample;

import android.content.Intent;
import android.os.Bundle;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

  private RecyclerView recyclerView;
  private NoteAdapter adapter;
  private ArrayList<NoteModel> notes = new ArrayList<>();
  private ActivityResultLauncher<Intent> activityResultLauncher
      = registerForActivityResult(
      new ActivityResultContracts.StartActivityForResult(),
      result -> {
        if (result.getResultCode() == RESULT_OK) {
          loadFromDatabase();
        }
      }
  );

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    recyclerView = (RecyclerView) findViewById(R.id.recycler_view);

    recyclerView.addItemDecoration(new DividerItemDecoration(this,
        DividerItemDecoration.VERTICAL)); // разделители элементов списка

    recyclerView.setLayoutManager(new LinearLayoutManager(this));
    adapter = new NoteAdapter();
    recyclerView.setAdapter(adapter);

    findViewById(R.id.btn_add).setOnClickListener(v -> {
      launchCreateNoteScreen();
    });

    loadFromDatabase();
  }

  private void launchCreateNoteScreen() {
    Intent intent = new Intent(MainActivity.this, CreateNoteActivity.class);

    activityResultLauncher.launch(intent);
  }

  /**
   * Загрузка заметок из базы даннх
   */
  private void loadFromDatabase() {
    DatabaseHelper dbHelper = new DatabaseHelper(this);
    notes = dbHelper.getAllNotes();
    adapter.updateNotes(notes);
  }
}

File: app/src/main/java/com/eh/recyclerlistsample/MainActivity.kt
package com.eh.recyclerlistsample

import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import android.content.Intent
import androidx.activity.result.contract.ActivityResultContracts.StartActivityForResult
import android.os.Bundle
import android.view.View
import androidx.activity.result.ActivityResult
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.eh.recyclerlistsample.R.id
import com.eh.recyclerlistsample.R.layout

class MainActivity : AppCompatActivity() {

  private lateinit var recyclerView: RecyclerView
  private lateinit var adapter: NoteAdapter

  private val activityResultLauncher = registerForActivityResult(
    StartActivityForResult()
  ) { result: ActivityResult ->
    if (result.resultCode == RESULT_OK) {
      loadFromDatabase()
    }
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(layout.activity_main)
    recyclerView = findViewById<View>(id.recycler_view) as RecyclerView
    recyclerView.addItemDecoration(
      DividerItemDecoration(
        this,
        DividerItemDecoration.VERTICAL
      )
    ) // разделители элементов списка
    recyclerView.layoutManager = LinearLayoutManager(this)
    adapter = NoteAdapter()
    recyclerView.adapter = adapter
    findViewById<View>(id.btn_add).setOnClickListener { launchCreateNoteScreen() }
    loadFromDatabase()
  }

  private fun launchCreateNoteScreen() {
    val intent = Intent(this@MainActivity, CreateNoteActivity::class.java)
    activityResultLauncher.launch(intent)
  }

  /**
   * Загрузка заметок из базы даннх
   */
  private fun loadFromDatabase() {
    val dbHelper = BaseApplication.instance.database
    val notes = dbHelper.notesDao().getAll()
    adapter.updateNotes(notes.map {
      it.toModel()
    })
  }
}